simuleval/agents/agent.py

```diff
@@ -4,54 +4,163 @@
-            return SpeechStates
-        else:
-            raise NotImplementedError
+class GenericAgent:
+    """
+    Generic Agent class.
+    """
 
-    def segment_to_units(self, segment, states):
-        return [segment]
+    source_type = None
+    target_type = None
 
-    def units_to_segment(self, unit_queue, states):
-        return unit_queue.pop()
+    def __init__(self, args: Namespace) -> None:
+        self.args = args
+        assert self.source_type
+        assert self.target_type
 
-    def update_states_read(self, states):
-        pass
+        self.states = self.build_states()
+        self.reset()
 
-    def update_states_write(self, states):
-        pass
+    def build_states(self) -> AgentStates:
+        """
+        Build states instance for agent
 
-    def build_states(self, args, client, sentence_id):
-        # Initialize states here, for example add customized entry to states
-        # This funcion will be caused at begining of every new sentence
-        states = self.states_type(args)(args, client, sentence_id, self)
-        self.initialize_states(states)
-        return states
+        Returns:
+            AgentStates: agent states
+        """
+        return AgentStates()
 
-    def initialize_states(self, states):
-        pass
+    def reset(self) -> None:
+        """
+        Reset agent, called every time when a new sentence coming in.
+        """
+        self.states.reset()
+
+    def policy(self) -> Action:
+        """
+        The policy to make decision every time
+        when the system has new input.
+        The function has to return an Action instance
+
+        Returns:
+            Action: The actions to make at certain point.
+
+        .. note:
+
+            WriteAction means that the system has a prediction.
+            ReadAction means that the system needs more source.
+        """
+        assert NotImplementedError
+
+    def push(self, source_segment: Segment) -> None:
+        """
+        The function to process the incoming information.
+
+        Args:
+            source_info (dict): incoming information dictionary
+        """
+        self.states.update_source(source_segment)
+
+    def pop(self) -> Segment:
+        """
+        The function to generate system output.
+        By default, it first runs policy,
+        and than returns the output segment.
+        If the policy decide to read,
+        it will return an empty segment.
+
+
+        Returns:
+            Segment: segment to return.
+        """
+
+        action = self.policy()
+        if not isinstance(action, Action):
+            raise RuntimeError(
+                f"The return value of {self.policy.__qualname__} is not a {Action.__qualname__} instance"
+            )
+        if action.is_read():
+            return EmptySegment()
+        else:
+            segment = SEGMENT_TYPE_DICT[self.target_type](
+                index=0, content=action.content, finished=action.finished
+            )
```

Comment:
Just a thought -- the policy method is currently responsible for both determining whether to read or write, as well as making the prediction. (So here we're essentially just creating a segment from the prediction created in the policy and returned in the action object). I'm wondering if it makes sense to separate the two. In other words, self.policy() just determines whether we should read or write, and then we have another method (self.write()) which returns a segment and we'd replace this line here with that.

Not super critical -- but I wonder if there could be benefits. For example, we could reuse the same wait k policy agent, but then subclass it with different ways of generating the target.

Reply:
(Not suggesting a change, just food for thought)

Reply:
Good suggestion! I used to have that design but it seems make thing complicated. Maybe the previous design is not good. We can revisit this idea later